[JSC] WASM IPInt SIMD: implement extract opcodes

dhecht opened the pull request:
#### 0ce8441cd8bf9476478643007d540981a72bb079
<pre>
[JSC] WASM IPInt SIMD: implement extract opcodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=298365">https://bugs.webkit.org/show_bug.cgi?id=298365</a>
<a href="https://rdar.apple.com/159810092">rdar://159810092</a>

Reviewed by Yusuke Suzuki.

Implement extract opcode variants using emulation by loading
the appropriate lane from the stack.

I didn&apos;t find a test that systematically tests each instruction
in isolation, so adding test:
JSTests/wasm/stress/simd-instructions-extract-lanes.js.

Canonical link: <a href="https://commits.webkit.org/299593@main">https://commits.webkit.org/299593@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dbf2b050b36280a4f839a4827ec24901557f0b04

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/119521 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39210 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29865 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/125791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/71593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121395 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/39909 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47791 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/125791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/71593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122470 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/39909 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/107168 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/125791 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/39909 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/25273 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/69440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/111646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/39909 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/25466 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/128767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/118037 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/46441 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/128767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/46806 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/103362 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/128767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/46303 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/52009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/146735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/45767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/146735 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/49118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/47455 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -4094,46 +4100,112 @@ unimplementedInstruction(_simd_f32x4_splat)
+    loadb ImmLaneIdxOffset[PC], t0
+    andi ImmLaneIdx16Mask, t0
+    loadb [sp, t0], t0
+    sxb2i t0, t0
```
Let's use `loadbsi [sp, t0], t0`.

dhecht replied:
Done, thanks.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -4094,46 +4100,112 @@ unimplementedInstruction(_simd_f32x4_splat)
+    andi ImmLaneIdx8Mask, t0
+    lshifti 1, t0
+    loadh [sp, t0], t0
+    sxh2i t0, t0
```
Let's use `loadhsi [sp, t0, 2], t0`.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -4094,46 +4100,112 @@ unimplementedInstruction(_simd_f32x4_splat)
+    loadb ImmLaneIdxOffset[PC], t0
+    andi ImmLaneIdx8Mask, t0
+    lshifti 1, t0
+    loadh [sp, t0], t0
```
Let's use `loadh [sp, t0, 2], t0`.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -4094,46 +4100,112 @@ unimplementedInstruction(_simd_f32x4_splat)
+    loadb ImmLaneIdxOffset[PC], t0
+    andi ImmLaneIdx4Mask, t0
+    lshifti 2, t0
+    loadi [sp, t0], t0
```
Let's use `loadi [sp, t0, 4], t0`

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -4094,46 +4100,112 @@ unimplementedInstruction(_simd_f32x4_splat)
+    loadb ImmLaneIdxOffset[PC], t0
+    andi ImmLaneIdx2Mask, t0
+    lshifti 3, t0
+    loadq [sp, t0], t0
```
Let's use `loadq [sp, t0, 8] t0`

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -4094,46 +4100,112 @@ unimplementedInstruction(_simd_f32x4_splat)
+    loadb ImmLaneIdxOffset[PC], t0
+    andi ImmLaneIdx4Mask, t0
+    lshifti 2, t0
+    loadf [sp, t0], ft0
```
Let's use `loadf [sp, t0, 4], ft0`.

Constellation commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -4094,46 +4100,112 @@ unimplementedInstruction(_simd_f32x4_splat)
+    loadb ImmLaneIdxOffset[PC], t0
+    andi ImmLaneIdx2Mask, t0
+    lshifti 3, t0
+    loadd [sp, t0], ft0
```
Let's use `loadd [sp, t0, 8], ft0`.

Constellation reviewed (approved):
r=me with comments.

dhecht committed:
[JSC] WASM IPInt SIMD: implement extract opcodes
https://bugs.webkit.org/show_bug.cgi?id=298365
rdar://159810092

Reviewed by Yusuke Suzuki.

Implement extract opcode variants using emulation by loading
the appropriate lane from the stack.

I didn't find a test that systematically tests each instruction
in isolation, so adding test:
JSTests/wasm/stress/simd-instructions-extract-lanes.js.

Canonical link: https://commits.webkit.org/299593@main
